<a href="https://colab.research.google.com/github/ArtiPhadke/Arti_INFO5731_Spring2023/blob/main/In_class_exercise_04_03282023_ArtiPhadke.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **The fourth in-class-exercise (40 points in total, 03/28/2022)**

Arti Atul Phadke

In class exercise 4 

Question description: Please use the text corpus you collected in your last in-class-exercise for this exercise. Perform the following tasks:

## (1) (10 points) Generate K topics by using LDA, the number of topics K should be decided by the coherence score, then summarize what are the topics. You may refer the code here: 

https://www.machinelearningplus.com/nlp/topic-modeling-gensim-python/

In [ ]:
# Write your code here
# Importing the required libraries
import pandas as pd

# Load data from CSV to dataframe - 2 columns - show_id & show description
movies_lda = pd.read_csv('netflix_titles.csv', usecols=['show_id', 'description'])

# Show first few rows from dataframe
movies_lda.head()

show_id                                        description
0      s1  As her father nears the end of his life, filmm...
1      s2  After crossing paths at a party, a Cape Town t...
2      s3  To protect his family from a powerful drug lor...
3      s4  Feuds, flirtations and toilet talk go down amo...
4      s5  In a city of coaching centers known to train I...

In [ ]:
# Load the regular expression library
import re

# Remove punctuation
movies_lda['description_processed'] = \
movies_lda['description'].map(lambda x: re.sub('[,\.!?]', '', x))

# Convert the titles to lowercase
movies_lda['description_processed'] = \
movies_lda['description_processed'].map(lambda x: x.lower())

# Show first few rows from dataframe
movies_lda['description_processed'].head()

0    as her father nears the end of his life filmma...
1    after crossing paths at a party a cape town te...
2    to protect his family from a powerful drug lor...
3    feuds flirtations and toilet talk go down amon...
4    in a city of coaching centers known to train i...
Name: description_processed, dtype: object

In [ ]:
# Import required gensim libraries
import gensim
from gensim.utils import simple_preprocess
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords

# List and remove stopwords
stop_words = stopwords.words('english')
stop_words.extend(['from', 'subject', 're', 'edu', 'use'])

# Define functions to remove the stopwords
def sent_to_words(sentences):
    for sentence in sentences:
        yield(gensim.utils.simple_preprocess(str(sentence), deacc=True))
def remove_stopwords(texts):
    return [[word for word in simple_preprocess(str(doc)) 
             if word not in stop_words] for doc in texts]

# Cleaned 'description' column is converted into list
data = movies_lda.description_processed.values.tolist()
data_words = list(sent_to_words(data))

# Use the function to remove stop words
data_words = remove_stopwords(data_words)

# Print the final list of words with stopwords removed, for the first 5 descriptions
print(data_words[:1][0][:30])
print(data_words[1:2][0][:30])
print(data_words[2:3][0][:30])
print(data_words[3:4][0][:30])
print(data_words[4:5][0][:30])


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


['father', 'nears', 'end', 'life', 'filmmaker', 'kirsten', 'johnson', 'stages', 'death', 'inventive', 'comical', 'ways', 'help', 'face', 'inevitable']
['crossing', 'paths', 'party', 'cape', 'town', 'teen', 'sets', 'prove', 'whether', 'private', 'school', 'swimming', 'star', 'sister', 'abducted', 'birth']
['protect', 'family', 'powerful', 'drug', 'lord', 'skilled', 'thief', 'mehdi', 'expert', 'team', 'robbers', 'pulled', 'violent', 'deadly', 'turf', 'war']
['feuds', 'flirtations', 'toilet', 'talk', 'go', 'among', 'incarcerated', 'women', 'orleans', 'justice', 'center', 'new', 'orleans', 'gritty', 'reality', 'series']
['city', 'coaching', 'centers', 'known', 'train', 'india', 'finest', 'collegiate', 'minds', 'earnest', 'unexceptional', 'student', 'friends', 'navigate', 'campus', 'life']


In [ ]:
# Import the required libraries
import gensim.corpora as corpora
# Create Dictionary using cleaned list of data
id2word = corpora.Dictionary(data_words)
# Create Corpus to be used for LDA model
texts = data_words
# Term Document Frequency 
corpus = [id2word.doc2bow(text) for text in texts]
# View the TFIDF for first 5 descriptions
print(corpus[:1][0][:30])
print(corpus[1:2][0][:30])
print(corpus[2:3][0][:30])
print(corpus[3:4][0][:30])
print(corpus[4:5][0][:30])

[(0, 1), (1, 1), (2, 1), (3, 1), (4, 1), (5, 1), (6, 1), (7, 1), (8, 1), (9, 1), (10, 1), (11, 1), (12, 1), (13, 1), (14, 1)]
[(15, 1), (16, 1), (17, 1), (18, 1), (19, 1), (20, 1), (21, 1), (22, 1), (23, 1), (24, 1), (25, 1), (26, 1), (27, 1), (28, 1), (29, 1), (30, 1)]
[(31, 1), (32, 1), (33, 1), (34, 1), (35, 1), (36, 1), (37, 1), (38, 1), (39, 1), (40, 1), (41, 1), (42, 1), (43, 1), (44, 1), (45, 1), (46, 1)]
[(47, 1), (48, 1), (49, 1), (50, 1), (51, 1), (52, 1), (53, 1), (54, 1), (55, 1), (56, 2), (57, 1), (58, 1), (59, 1), (60, 1), (61, 1)]
[(11, 1), (62, 1), (63, 1), (64, 1), (65, 1), (66, 1), (67, 1), (68, 1), (69, 1), (70, 1), (71, 1), (72, 1), (73, 1), (74, 1), (75, 1), (76, 1)]


In [ ]:
# Import the required libraries
from gensim import corpora

# Create a dictionary with the corpus
corpus = data_words
dictionary = corpora.Dictionary(corpus)

# Convert the corpus into a bag of words
bow = [dictionary.doc2bow(text) for text in corpus]

In [ ]:
# Import the required libraries
from gensim.models import LsiModel
from gensim.models.coherencemodel import CoherenceModel

# Coherence score in topic modeling to measure how interpretable the topics are to humans.
# Find the coherence score with a different number of topics, between 2 to 10
for i in range(2,11):
    lsi = LsiModel(bow, num_topics=i, id2word=dictionary)

    # Run the CoherenceModel using LsiModel
    coherence_model = CoherenceModel(model=lsi, texts=data_words, dictionary=dictionary, coherence='c_v')
    coherence_score = coherence_model.get_coherence()
    print('Coherence score with {} clusters: {}'.format(i, coherence_score))

Coherence score with 2 clusters: 0.3183289864323845
Coherence score with 3 clusters: 0.3065190011578253
Coherence score with 4 clusters: 0.3042466137490799
Coherence score with 5 clusters: 0.265031978166268
Coherence score with 6 clusters: 0.25995509740640094
Coherence score with 7 clusters: 0.24994034846743293
Coherence score with 8 clusters: 0.2654729413003503
Coherence score with 9 clusters: 0.25721988453759637
Coherence score with 10 clusters: 0.2437106370744849


In [ ]:
# Import the required libraries
from pprint import pprint
from gensim.models.ldamulticore import LdaMulticore

# Choose the number of topics to be = 2, where maximum Coherence score is achieved = 0.3183
num_topics = 2
# Build LDA model
lda_model = gensim.models.LdaMulticore(corpus=data_words, id2word=id2word, num_topics=num_topics)

# Print the Keyword in the 10 topics
pprint(lda_model.print_topics())
doc_lda = lda_model[corpus]

In [ ]:
!pip install pyLDAvis

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 30.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 298.0/298.0 KB 23.5 MB/s eta 0:00:00
  Attempting uninstall: joblib
    Found existing installation: joblib 1.1.1
    Uninstalling joblib-1.1.1:
      Successfully uninstalled joblib-1.1.1
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
pandas-profiling 3.2.0 requires joblib~=1.1.0, but you have joblib 1.2.0 which is incompatible.


In [ ]:
!pip install --upgrade ipykernel

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 150.0/150.0 KB 7.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 49.2 MB/s eta 0:00:00
  Attempting uninstall: ipykernel
    Found existing installation: ipykernel 5.5.6
    Uninstalling ipykernel-5.5.6:
      Successfully uninstalled ipykernel-5.5.6
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires ipykernel~=5.5.6, but you have ipykernel 6.22.0 which is incompatible.


In [ ]:
# Import pyLDAvis.gensim model libraries
import pyLDAvis
import pyLDAvis.gensim_models as gensimvis
pyLDAvis.enable_notebook()

# Prepare the model
LDAvis_prepared = pyLDAvis.gensim_models.prepare(lda_model, corpus, id2word)

# Display the output for the first two topics and show top 30 relevant terms in both the topics
pyLDAvis.display(LDAvis_prepared)

/usr/local/lib/python3.9/dist-packages/pyLDAvis/_prepare.py:243: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  default_term_info = default_term_info.sort_values(


## (2) (10 points) Generate K topics by using LSA, the number of topics K should be decided by the coherence score, then summarize what are the topics. You may refer the code here:

https://www.datacamp.com/community/tutorials/discovering-hidden-topics-python

In [ ]:
# Write your code here

# Import the libraries
import pandas as pd

# Load data from csv file, using the columns 'show_id', 'description'
df = pd.read_csv('netflix_titles.csv', usecols=['show_id', 'description'])
df.head()

show_id                                        description
0      s1  As her father nears the end of his life, filmm...
1      s2  After crossing paths at a party, a Cape Town t...
2      s3  To protect his family from a powerful drug lor...
3      s4  Feuds, flirtations and toilet talk go down amo...
4      s5  In a city of coaching centers known to train I...

In [ ]:
# Check and confirm there is no Null value in columns selected
df['description'].isna().value_counts()

False    8807
Name: description, dtype: int64

In [ ]:
!pip install gensim

In [ ]:
!pip install --upgrade numpy

In [ ]:
# Import the libraries from gensim for preprocessing of data
from gensim.parsing.preprocessing import remove_stopwords, strip_punctuation , preprocess_string, strip_short, stem_text

# Preprocess given text
def preprocess(text):
    
    # Clean text based on given filters
    CUSTOM_FILTERS = [lambda x: x.lower(), 
                                remove_stopwords, 
                                strip_punctuation, 
                                strip_short, 
                                stem_text]
    text = preprocess_string(text, CUSTOM_FILTERS)
    
    return text

# Apply function to all the data description column
df['Text (Clean)'] = df['description'].apply(lambda x: preprocess(x))

In [ ]:
df.head()

show_id                                        description  \
0      s1  As her father nears the end of his life, filmm...   
1      s2  After crossing paths at a party, a Cape Town t...   
2      s3  To protect his family from a powerful drug lor...   
3      s4  Feuds, flirtations and toilet talk go down amo...   
4      s5  In a city of coaching centers known to train I...   

                                        Text (Clean)  
0  [father, near, end, life, filmmak, kirsten, jo...  
1  [cross, path, parti, cape, town, teen, set, pr...  
2  [protect, famili, power, drug, lord, skill, th...  
3  [feud, flirtat, toilet, talk, incarcer, women,...  
4  [citi, coach, center, known, train, india’, fi...

In [ ]:
# Import required libraries
from gensim import corpora

# Create a dictionary with the corpus, which contains the cleaned data
corpus = df['Text (Clean)']
dictionary = corpora.Dictionary(corpus)

# Convert corpus into a bag of words 
bow = [dictionary.doc2bow(text) for text in corpus]

In [ ]:
# Import required libraries
from gensim.models import LsiModel
from gensim.models.coherencemodel import CoherenceModel

# Coherence score in topic modeling to measure how interpretable the topics are to humans.
# find the coherence score with a different number of topics from 2 to 10
for i in range(2,11):
    lsi = LsiModel(bow, num_topics=i, id2word=dictionary)
    coherence_model = CoherenceModel(model=lsi, texts=df['Text (Clean)'], dictionary=dictionary, coherence='c_v')
    coherence_score = coherence_model.get_coherence()
    print('Coherence score with {} clusters: {}'.format(i, coherence_score))

Coherence score with 2 clusters: 0.2518260491635272
Coherence score with 3 clusters: 0.31217603759955453
Coherence score with 4 clusters: 0.2820376405127971
Coherence score with 5 clusters: 0.2964000694749021
Coherence score with 6 clusters: 0.2764828152027305
Coherence score with 7 clusters: 0.27090800926535186
Coherence score with 8 clusters: 0.29145344808956164
Coherence score with 9 clusters: 0.2653853147745181
Coherence score with 10 clusters: 0.27667960344014825


In [ ]:
# Perform SVD on the bag of words with the LsiModel to extract 3 topics
# Since the maximum coherence score is obtained for number of clusters = 3 and create Lsimodel
lsi = LsiModel(bow, num_topics=3, id2word=dictionary)

In [ ]:
# Find the 10 words with the srongest association to the derived topics
for topic_num, words in lsi.print_topics(num_words=10):
    print('Words in {}: {}.'.format(topic_num, words))

Words in 0: 0.326*"life" + 0.275*"young" + 0.269*"new" + 0.231*"famili" + 0.227*"love" + 0.197*"friend" + 0.183*"man" + 0.171*"woman" + 0.151*"world" + 0.151*"year".
Words in 1: -0.563*"young" + 0.476*"new" + -0.309*"man" + 0.294*"life" + -0.252*"woman" + 0.177*"friend" + -0.163*"love" + 0.139*"world" + -0.117*"famili" + 0.107*"school".
Words in 2: 0.802*"life" + -0.356*"new" + -0.234*"friend" + 0.186*"love" + -0.109*"help" + -0.097*"world" + -0.083*"school" + -0.078*"home" + -0.073*"famili" + -0.069*"young".


In [ ]:
# Find the scores given between the review and each topic
corpus_lsi = lsi[bow]
score1 = []
score2 = []
for doc in corpus_lsi:
    score1.append(round(doc[0][1],2))
    score2.append(round(doc[1][1],2))

# Create data frame that shows scores assigned for both topics for each review
df_topic = pd.DataFrame()
df_topic['Text'] = df['description']
df_topic['Topic 0 score'] = score1
df_topic['Topic 1 score'] = score2
df_topic['Topic']= df_topic[['Topic 0 score', 'Topic 1 score']].apply(lambda x: x.argmax(), axis=1)
df_topic.head(1)

Text  Topic 0 score  \
0  As her father nears the end of his life, filmm...           0.79   

   Topic 1 score  Topic  
0           0.26      0

In [ ]:
# Find a sample review from each topic
df_topic0 = df_topic[df_topic['Topic'] == 0]
df_topic1 = df_topic[df_topic['Topic']==1]
print('Sample text from topic 0:\n {}'.format(df_topic0.sample(1, random_state=2)['Text'].values))
print('\nSample text from topic 1:\n {}'.format(df_topic1.sample(1, random_state=2)['Text'].values))

Sample text from topic 0:
 ["Wild artistic inspiration and emotional turmoil fill painter Vincent van Gogh's last years as he struggles to bring his unique vision to the world."]

Sample text from topic 1:
 ['When the Hong Kong government enacts a ban on smoking cigarettes indoors, the new law drives hard-core smokers outside, facilitating unlikely connections.']


## (3) (10 points) Generate K topics by using  lda2vec, the number of topics K should be decided by the coherence score, then summarize what are the topics. You may refer the code here:

https://nbviewer.org/github/cemoody/lda2vec/blob/master/examples/twenty_newsgroups/lda2vec/lda2vec.ipynb

In [ ]:
!pip install lda2vec
!pip install preprocessing
!pip install Corpus

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.1/88.1 KB 8.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for Corpus: filename=Corpus-0.4.2-py3-none-any.whl size=88795 sha256=66fdcd72282a0fef565d2585ee83e21a38ecd1b56c53627ad2c79f10ae3b9748
  Stored in directory: /root/.cache/pip/wheels/73/b1/96/d08538d62b0ad5b467fe9080873b4dbfe011044d687656939c
Successfully built Corpus


In [ ]:
# Write your code here
# Import required libraries
import matplotlib.pyplot as plt
import numpy as np

import pyLDAvis
pyLDAvis.enable_notebook()


In [ ]:

# Import the required libraries
import pandas as pd

# Load data from CSV to dataframe - 2 columns - show_id & show description
movies_lda = pd.read_csv('netflix_titles.csv', usecols=['show_id', 'description'])

# Show first few rows from dataframe
movies_lda.head()
# Load the regular expression library
import re

# Remove punctuation
movies_lda['description_processed'] = \
movies_lda['description'].map(lambda x: re.sub('[,\.!?]', '', x))

# Convert the titles to lowercase
movies_lda['description_processed'] = \
movies_lda['description_processed'].map(lambda x: x.lower())

# Show first few rows from dataframe
movies_lda['description_processed'].head()

# Import required gensim libraries
import gensim
from gensim.utils import simple_preprocess
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords

# List and remove stopwords
stop_words = stopwords.words('english')
stop_words.extend(['from', 'subject', 're', 'edu', 'use'])

# Define functions to remove the stopwords
def sent_to_words(sentences):
    for sentence in sentences:
        yield(gensim.utils.simple_preprocess(str(sentence), deacc=True))
def remove_stopwords(texts):
    return [[word for word in simple_preprocess(str(doc)) 
             if word not in stop_words] for doc in texts]

# Cleaned 'description' column is converted into list
data = movies_lda.description_processed.values.tolist()
data_words = list(sent_to_words(data))

# Use the function to remove stop words
data_words = remove_stopwords(data_words)

# Print the final list of words with stopwords removed, for the first 5 descriptions
print(data_words[:1][0][:30])
print(data_words[1:2][0][:30])
print(data_words[2:3][0][:30])
print(data_words[3:4][0][:30])
print(data_words[4:5][0][:30])

# Import the required libraries
import gensim.corpora as corpora
# Create Dictionary using cleaned list of data
id2word = corpora.Dictionary(data_words)
# Create Corpus to be used for LDA model
texts = data_words
# Term Document Frequency 
corpus = [id2word.doc2bow(text) for text in texts]
# View the TFIDF for first 5 descriptions
print(corpus[:1][0][:30])
print(corpus[1:2][0][:30])
print(corpus[2:3][0][:30])
print(corpus[3:4][0][:30])
print(corpus[4:5][0][:30])

# Import the required libraries
from gensim import corpora

# Create a dictionary with the corpus
corpus = data_words
dictionary = corpora.Dictionary(corpus)

# Convert the corpus into a bag of words
bow = [dictionary.doc2bow(text) for text in corpus]

# Import the required libraries
from gensim.models import LsiModel
from gensim.models.coherencemodel import CoherenceModel

# Coherence score in topic modeling to measure how interpretable the topics are to humans.
# Find the coherence score with a different number of topics, between 2 to 10
for i in range(2,11):
    lsi = LsiModel(bow, num_topics=i, id2word=dictionary)

    # Run the CoherenceModel using LsiModel
    coherence_model = CoherenceModel(model=lsi, texts=data_words, dictionary=dictionary, coherence='c_v')
    coherence_score = coherence_model.get_coherence()
    print('Coherence score with {} clusters: {}'.format(i, coherence_score))


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
['father', 'nears', 'end', 'life', 'filmmaker', 'kirsten', 'johnson', 'stages', 'death', 'inventive', 'comical', 'ways', 'help', 'face', 'inevitable']
['crossing', 'paths', 'party', 'cape', 'town', 'teen', 'sets', 'prove', 'whether', 'private', 'school', 'swimming', 'star', 'sister', 'abducted', 'birth']
['protect', 'family', 'powerful', 'drug', 'lord', 'skilled', 'thief', 'mehdi', 'expert', 'team', 'robbers', 'pulled', 'violent', 'deadly', 'turf', 'war']
['feuds', 'flirtations', 'toilet', 'talk', 'go', 'among', 'incarcerated', 'women', 'orleans', 'justice', 'center', 'new', 'orleans', 'gritty', 'reality', 'series']
['city', 'coaching', 'centers', 'known', 'train', 'india', 'finest', 'collegiate', 'minds', 'earnest', 'unexceptional', 'student', 'friends', 'navigate', 'campus', 'life']
[(0, 1), (1, 1), (2, 1), (3, 1), (4, 1), (5, 1), (6, 1), (7, 1), (8, 1), (9, 1), (10,

In [ ]:
# Import the required libraries for lda2vec
from gensim.models.ldamulticore import LdaMulticore

# Choose the number of topics to be = 5, where maximum Coherence score is achieved = 0.3183
num_topics = 5
# Build LDA model
lda_model = gensim.models.LdaMulticore(corpus=data_words, id2word=id2word, num_topics=num_topics)

# Print the Keyword in the 10 topics
pprint(lda_model.print_topics())
doc_lda = lda_model[corpus]

## (4) (10 points) Generate K topics by using BERTopic, the number of topics K should be decided by the coherence score, then summarize what are the topics. You may refer the code here: 

https://colab.research.google.com/drive/1FieRA9fLdkQEGDIMYl0I3MCjSUKVF8C-?usp=sharing

In [ ]:
# Write your code here

# Import the required libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Load the data from csv file
df_movies = pd.read_csv('netflix_titles.csv')

# Show the first few rows from the csv file
df_movies.head()


/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [ ]:
from gensim.parsing.preprocessing import remove_stopwords, strip_punctuation , preprocess_string, strip_short, stem_text

# Preprocess the given data
def preprocess(text):
    
    # clean text based on given filters
    CUSTOM_FILTERS = [lambda x: x.lower(), 
                                remove_stopwords, 
                                strip_punctuation, 
                                strip_short, 
                                stem_text]
    text = preprocess_string(text, CUSTOM_FILTERS)
    
    return text

# Apply the function to all reviews 
df_movies['clean_text'] = df_movies['description'].apply(lambda x: preprocess(x))

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
!pip install bertopic

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.1/17.1 MB 51.6 MB/s eta 0:00:00
  Attempting uninstall: numpy
    Found existing installation: numpy 1.24.2
    Uninstalling numpy-1.24.2:
      Successfully uninstalled numpy-1.24.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
pandas-profiling 3.2.0 requires joblib~=1.1.0, but you have joblib 1.2.0 which is incompatible.
google-colab 1.0.0 requires ipykernel~=5.5.6, but you have ipykernel 6.22.0 which is incompatible.


In [ ]:
# Import the required libraries
from bertopic import BERTopic

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
# Clean the data
desc = df_movies.clean_text.to_list()

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
# Create the BERTopic model
topic_model = BERTopic(language="english")
topics, probs = topic_model.fit_transform(df_movies.description)

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
# Visualize the results
topic_model.visualize_topics()

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
# Import the required libraries
from gensim import corpora

# Create a dictionary with the corpus
corpus = df_movies['clean_text']
dictionary = corpora.Dictionary(corpus)

# Convert corpus into a bag of words
bow = [dictionary.doc2bow(text) for text in corpus]

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
# Import the required libraries
from gensim.models import LsiModel
from gensim.models.coherencemodel import CoherenceModel

# Coherence score in topic modeling to measure how interpretable the topics are to humans.
# find the coherence score with a different number of topics
for i in range(2,11):
    lsi = LsiModel(bow, num_topics=i, id2word=dictionary)
    coherence_model = CoherenceModel(model=lsi, texts=df['Text (Clean)'], dictionary=dictionary, coherence='c_v')
    coherence_score = coherence_model.get_coherence()
    print('Coherence score with {} clusters: {}'.format(i, coherence_score))

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.

/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning:

Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.



Coherence score with 2 clusters: 0.30290104821950004


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning:

Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.



Coherence score with 3 clusters: 0.31217603759955453


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning:

Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.



Coherence score with 4 clusters: 0.25824166352928296


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning:

Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.



Coherence score with 5 clusters: 0.2866078659013954


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning:

Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.



Coherence score with 6 clusters: 0.2927314071787635


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning:

Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.



Coherence score with 7 clusters: 0.255018427475257


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning:

Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.



Coherence score with 8 clusters: 0.26829466606434277


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning:

Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.



Coherence score with 9 clusters: 0.24851480149405825


/usr/local/lib/python3.9/dist-packages/gensim/models/lsimodel.py:963: DeprecationWarning:

Please use `csc_matvecs` from the `scipy.sparse` namespace, the `scipy.sparse.sparsetools` namespace is deprecated.



Coherence score with 10 clusters: 0.26646594230951687


In [ ]:
# Visualize the results for no. of topics = 3
topic_model.visualize_barchart(top_n_topics=3, n_words = 10, width = 350, height = 350)

/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



## (5) (10 extra points) Compare the results generated by the four topic modeling algorithms, which one is better? You should explain the reasons in details.

In [ ]:
# Write your answer here (no code needed for this question)


The results generated by the above methods can be compared as follows:

1) LDA model: The corpus generated for first 5 topics is as follows. The best coherence score is obtained for no. of topics = 2
['father', 'nears', 'end', 'life', 'filmmaker', 'kirsten', 'johnson', 'stages', 'death', 'inventive', 'comical', 'ways', 'help', 'face', 'inevitable']
['crossing', 'paths', 'party', 'cape', 'town', 'teen', 'sets', 'prove', 'whether', 'private', 'school', 'swimming', 'star', 'sister', 'abducted', 'birth']
['protect', 'family', 'powerful', 'drug', 'lord', 'skilled', 'thief', 'mehdi', 'expert', 'team', 'robbers', 'pulled', 'violent', 'deadly', 'turf', 'war']
['feuds', 'flirtations', 'toilet', 'talk', 'go', 'among', 'incarcerated', 'women', 'orleans', 'justice', 'center', 'new', 'orleans', 'gritty', 'reality', 'series']
['city', 'coaching', 'centers', 'known', 'train', 'india', 'finest', 'collegiate', 'minds', 'earnest', 'unexceptional', 'student', 'friends', 'navigate', 'campus', 'life']

After this, the top 10 terms related to first 2 topics are as follows:
topic 1: life, new, family, young, two, woman, series, man, love, world
topic 2: young, world, life, man, new, documentary, woman, two, one, find

2) LSA model: The best coherence score is obtained for no. of topics = 3. The terms chosen in every topic are as follows:

Words in 0: 0.326*"life" + 0.275*"young" + 0.269*"new" + 0.231*"famili" + 0.227*"love" + 0.197*"friend" + 0.183*"man" + 0.171*"woman" + 0.151*"world" + 0.151*"year".
Words in 1: -0.563*"young" + 0.476*"new" + -0.309*"man" + 0.294*"life" + -0.252*"woman" + 0.177*"friend" + -0.163*"love" + 0.139*"world" + -0.117*"famili" + 0.107*"school".
Words in 2: 0.802*"life" + -0.356*"new" + -0.234*"friend" + 0.186*"love" + -0.109*"help" + -0.097*"world" + -0.083*"school" + -0.078*"home" + -0.073*"famili" + -0.069*"young".

topic 1: life, young, new, famili, love, friend, man, woman, world, year
topic 2: young, new, man, life, woman, friend, love, world, famili, school
topic 3: life, new, friend, love, help, world, school, home, famili, young

3) LDA2VEC model: The best coherence score is obtained for no. of topics = 5. The terms chosen in top 2 topics are as follows:
topic 1: life, new, family, young, two, woman, series, man, love, world
topic 2: young, world, life, man, new, documentary, woman, two, one, find

4) BERT model: The best coherence score is obtained for no. of topics = 3. The terms chosen in the 3 topics are as follows:
topic 1: standup, comedian, comedy, special, comic, show, stage, and, more, humor
topic 2: detective, murder, killer, case, police, serial, cases, solve, detectives, crimes
topic 3: agent, terrorist, cia, mission, us, intelligence, attack, team, government, an


Here, the results generated by BERT model look like the best results. The topics chosen include a variety of different topics, like comedy, crime and terrorism. The variety of chosen topics is wider, selected terms by model related to every single topic are diverse and more closely related. 
Hence, in my understanding, BERT model looks like the best model.